File: ruby/lib/runapi/suno/resources/blend_lyrics.rb
# frozen_string_literal: true

module RunApi
  module Suno
    module Resources
      # Blends two caller-authored lyrics texts into a new lyrics result.
      class BlendLyrics
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/suno/blend_lyrics"
        RESPONSE_CLASS = Types::BlendLyricsResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedBlendLyricsResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_contract!(CONTRACT["blend-lyrics"], params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end
      end
    end
  end
end

File: python/src/runapi/suno/resources/blend_lyrics.py
"""Suno blend_lyrics resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import RequestOptions, Resource

from ..contract_gen import CONTRACT
from ..types import BlendLyricsResponse, CompletedBlendLyricsResponse


class BlendLyrics(Resource):
    """Blend two caller-authored lyrics texts."""

    ENDPOINT = "/api/v1/suno/blend_lyrics"
    RESPONSE_CLASS = BlendLyricsResponse
    COMPLETED_RESPONSE_CLASS = CompletedBlendLyricsResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["blend-lyrics"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)
